docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,257 @@
+* device.metadata.labels
+* device.metadata.name
+
+Making sure that the render of the device template is deterministic (i.e. not including timestamps or dates, etc..).
+
+#### Expected behavior
+
+When fleet.speec.template.metadata or fleet.spec.template.spec is updated, the Flet deployment controller
+starts rolling out that update to devices. This update can take long (depending on the upgrade strategy)
+
+When a template is applied to a specific device via the Device API:
+* The template variables get rendered (i.e. the labels at that point in time get used)
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes)
+* When later in time a device has labels updated, the device controller will queue the device to be re-evaulated and
+  the fleet template re-rendered. (this is the same workflow as when the labels selected by a fleet changes)
+
+### User Stories 
+
+#### Story 1: network access credentials
+As an IT administrator, I want to use site-specific network access credentials for devices in the same fleet, so that I do not have to deal with the complexity of maintaining one fleet per site.
+
+#### Story 2: site specific endpoints
+As an IT administrator, I want to provide site-specific server endpoints to devices that depend on LAN-configured edge servers.
+
+#### Story 3: site specific certificate bundles
+As an IT administrator, I want to provide site-specific certificate bundles to enable trust with the local site infrastructure.
+
+#### Story 4: injecting the device name into the configuration
+As an IT administrator, I want the name of my MicroShift instance in ACM to have the same cluster name as the underlying device's name or fingerprint, so I can easily correlate between the two.
+
+#### Story 5: site specific telemetry endpoints
+As an IT administrator, I want to provide site-specific/region-specific telemetry endpoints to devices.
+
+#### Story 6: dialing up verbosity for logging in my user application in specific devices
+As an IT administrator, I want to dial up the verbosity for logging on a specific device, so I can troubleshoot it.
+
+#### Story 7: region specific application service endpoints
+As an IT administrator, I want to provide region-specific service endpoints to my applications
+
+#### Story 8: region specific radio bands/regulations
+As a developer, I want to ensure that my devices automatically comply with local regulations when enrolled in specific regions. (i.e. configured with region-specific radio bands for WiFi, 5G, or other radio technologies)
+
+#### Story 9: region specific data sovereignty
+As a developer, I want to ensure that my devices connect to the specific regional endpoints to ensure compliance with data sovereignty.
+
+### Notes/Constraints/Caveats
+
+We would need to identify the design changes relative to templateVersions and
+if the relationship of those continues to be to the fleet or device level.
```

Comment:
This would be good to describe the ability to audit these templated variables in more detail. For example it should be possible to regenerate all of the device configurations in the past to validate correctness.

Reply:
Maybe not in the past, but the existing ones.

That would make sense as a good safety check.

We could also prevent modifying device labels via the device API when the rendering of the existing fleet is going to fail.

Reply:
I can see govt or insurance companies asking for how did this device get this spec on this day. Depending on the data retention policy of the user this should be possible in theory. Just stuff to think about.